tests/unit/fem/test_sparse_matrix.cpp

```diff
@@ -0,0 +1,204 @@
+   int dim = mesh.Dimension();
+
+   FiniteElementCollection *fec;
+   if (pb == Problem::Convection)
+   {
+      fec = new L2_FECollection(order, dim, basis);
+   }
+   else
+   {
+      fec = new H1_FECollection(order, dim, basis);
+   }
+   ParFiniteElementSpace fes(&pmesh, fec);
+
+   ParBilinearForm k_test(&fes);
+   ParBilinearForm k_ref(&fes);
+
+   ParFiniteElementSpace coeff_fes(&pmesh, fec);
+   ParGridFunction gf(&coeff_fes);
+
+   Coefficient *coeff = nullptr;
+   ConstantCoefficient rho(1.0);
+   VectorFunctionCoefficient velocity(dim, velocity_function);
+   switch (coeff_type)
+   {
+      case Coeff::Const:
+         coeff = new ConstantCoefficient(1.0);
+         break;
+      case Coeff::Grid:
+      {
+         FunctionCoefficient f_coeff(coeff_function);
+         gf.ProjectCoefficient(f_coeff);
+         coeff = new GridFunctionCoefficient(&gf);
+         break;
+      }
+      case Coeff::Quad:
+         coeff = new FunctionCoefficient(coeff_function);
+         break;
+   }
+
+   switch (pb)
+   {
+      case Problem::Mass:
+         k_ref.AddDomainIntegrator(new MassIntegrator(*coeff));
+         k_test.AddDomainIntegrator(new MassIntegrator(*coeff));
+         break;
+      case Problem::Convection:
+         k_ref.AddDomainIntegrator(new ConvectionIntegrator(velocity, -1.0));
+         k_ref.AddInteriorFaceIntegrator(
+            new TransposeIntegrator(new DGTraceIntegrator(rho, velocity, 1.0, -0.5)));
+         k_ref.AddBdrFaceIntegrator(
+            new TransposeIntegrator(new DGTraceIntegrator(rho, velocity, 1.0, -0.5)));
+         k_test.AddDomainIntegrator(new ConvectionIntegrator(velocity, -1.0));
+         k_test.AddInteriorFaceIntegrator(
+            new TransposeIntegrator(new DGTraceIntegrator(rho, velocity, 1.0, -0.5)));
+         k_test.AddBdrFaceIntegrator(
+            new TransposeIntegrator(new DGTraceIntegrator(rho, velocity, 1.0, -0.5)));
+         break;
+      case Problem::Diffusion:
+         k_ref.AddDomainIntegrator(new DiffusionIntegrator(*coeff));
+         k_test.AddDomainIntegrator(new DiffusionIntegrator(*coeff));
+         break;
+   }
+
+   if (keep_nbr_block) { k_ref.KeepNbrBlock(); }
+   k_ref.Assemble();
+   k_ref.Finalize();
+
+   k_test.SetAssemblyLevel(AssemblyLevel::FULL);
+   if (keep_nbr_block) { k_test.KeepNbrBlock(); }
+   k_test.Assemble();
+
+   const int sizeIn  = pb == Problem::Convection ?
+                       fes.GetVSize() + fes.GetFaceNbrVSize() :
+                       fes.GetVSize();
+   const int sizeOut = (pb == Problem::Convection && keep_nbr_block)?
+                       fes.GetVSize() + fes.GetFaceNbrVSize() :
+                       fes.GetVSize();
+   const int sizeEnd = fes.GetVSize();
+   Vector x(sizeIn), y_test(sizeOut), y_ref(sizeOut);
+   x.Randomize(1);
+
+   k_test.SpMat().Mult(x,y_test);
+   k_ref.SpMat().Mult(x,y_ref);
+
+   y_test -= y_ref;
+
+   Vector result(y_test.HostReadWrite(), sizeEnd);
+
+   REQUIRE(result.Norml2() < 1.e-12);
```

Comment:
This is a pretty loose tolerance for a single matvec.  Does the error get anywhere close to that?